cirq-google/cirq_google/ops/coupler_pulse.py

```diff
@@ -0,0 +1,123 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""An instance of FSimGate that works naturally on Google's Sycamore chip"""
+
+from typing import Optional, Tuple, TYPE_CHECKING
+
+import numpy as np
+
+from cirq import ops, protocols, value
+
+if TYPE_CHECKING:
+    import cirq
+
+
+_MIN_DURATION = value.Duration(nanos=0)
+_MAX_DURATION = value.Duration(nanos=100)
+
+
+class CouplerPulse(ops.gate_features.TwoQubitGate):
+    """Tunable pulse for entangling adjacent qubits.
+
+    For experimental usage only.
+
+    This operation sends a trapezoidal pulse to the coupler between two
+    adjacent qubits placed in resonance.
+
+    Note that this gate does not have a unitary matrix and must be
+    characterized by the user in order to determine its effects.
+
+    Args:
+            hold_time: Length of the 'plateau' part of the coupler trajectory.
+            coupling_MHz: Target qubit-qubit coupling reached at the plateau.
+            rise_time: Full width of the smoothstep rise/fall.
+            padding_time: Symmetric padding around the coupler pulse.
+    """
+
+    def __init__(
+        self,
+        hold_time: value.Duration,
+        coupling_mhz: float,
+        rise_time: Optional[value.Duration] = value.Duration(nanos=8),
+        padding_time: Optional[value.Duration] = value.Duration(nanos=2.5),
+    ):
+        """
+        Args:
+            hold_time: Length of the 'plateau' part of the coupler trajectory.
+            coupling_MHz: Target qubit-qubit coupling reached at the plateau.
+            rise_time: Full width of the smoothstep rise/fall.
+            padding_time: Symmetric padding around the coupler pulse.
+        """
+        # Verification
+        if hold_time > rise_time:
+            raise ValueError(
+                f'Full rise time {rise_time} must be longer '
+                'than hold_time {hold_time} for CouplerPulse'
+            )
+        if hold_time < _MIN_DURATION or hold_time > _MAX_DURATION:
+            raise ValueError(f'hold_time must be between {_MIN_DURATION} and {_MAX_DURATION}')
+        if padding_time < _MIN_DURATION or padding_time > _MAX_DURATION:
+            raise ValueError(f'padding_time must be between {_MIN_DURATION} and {_MAX_DURATION}')
+        if rise_time < _MIN_DURATION or rise_time > _MAX_DURATION:
+            raise ValueError(f'rise_time must be between {_MIN_DURATION} and {_MAX_DURATION}')
+
+        self.hold_time = hold_time
+        self.coupling_mhz = coupling_mhz
+        self.rise_time = rise_time or value.Duration(nanos=8)
+        self.padding_time = padding_time or value.Duration(nanos=2.5)
+
+    def num_qubits(self) -> int:
+        return 2
+
+    def _unitary_(self) -> np.ndarray:
+        return NotImplemented
```

Comment:
We have an approximate unitary implementation on the internal version of this gate. Do we want to add that to make it possible to simulate this out of the box?

Reply:
I am not sure about this either.  I saw that, but it uses scipy which is not included in cirq right now, which would be an additional dependency (or would need to be factored out).  It is also approximate, and I am not sure if it is a great idea to include, since our approximation could change or turn out to be incorrect (or differ from chip to chip).  I generally think it is better to leave it out for now.

Reply:
> but it uses scipy which is not included in cirq right now,
We do have scipy as our dependency in cirq-core...

It is weird to have a non-simulateable gate, so probably this will come as a feature request pretty quickly if this Gate gets picked up, but can be deferred.